File: python_leetcode_r1/1_TwoSum.py
from typing import List
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = {}
        for i, val in enumerate(nums):
            if target - val in map:
                return [map[target-val], i]
            map[val] = i
        return []

File: ts_leetcode_r1/1_TwoSum.ts
function twoSum(nums: number[], target: number): number[] {
    // exhaustive, start at index 0, check index 1, 2, 3 

    // is there earlier stop?

    // we are told not ordered

    // hashset store of nums
    // check (target - nums[i]) in hashset via single pass

    const numsToIndex = new Map<number, number>();

    for (let i = 0; i < nums.length; i++)   {
        const search = target - nums[i];
        if (numsToIndex.has(search)) {
            return [numsToIndex.get(search)!, i];
        }
        numsToIndex.set(nums[i], i);
    }

    return []; 
};